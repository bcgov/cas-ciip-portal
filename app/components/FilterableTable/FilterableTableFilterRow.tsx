import React, {useEffect, useState} from 'react';
import {Button} from 'react-bootstrap';
import {TableFilter, FilterArgs} from './Filters';

interface Props {
  filters: TableFilter[];
  filterArgs: FilterArgs;
  onSubmit: (searchData: Record<string, string | number | boolean>) => void;
}

const FilterableTableFilterRow: React.FunctionComponent<Props> = ({
  filters,
  filterArgs,
  onSubmit
}) => {
  const [searchFilters, setSearchFilters] = useState(filterArgs);
  useEffect(() => setSearchFilters(filterArgs), [filterArgs]); // reset the local state when the prop changes

  const handleFilterChange = (value, column) => {
    // using a state update with a callback ensures that we always have a reference to the latest searchFilters
    // especially when this handler is fired multiple times in quick sucession, without the component updating
    // which can happen when a single filter component handles multiple variables
    setSearchFilters((prevFilters) => ({
      ...prevFilters,
      [column]: value
    }));
  };

  const clearForm = () => {
    setSearchFilters({});
    onSubmit({});
  };

  const handleKeyDown: React.KeyboardEventHandler = (e) => {
    if (e.key === 'Enter') {
      onSubmit(searchFilters);
    }
  };

  return (
    <tr onKeyDown={handleKeyDown}>
      {filters.map((filter) => (
        <filter.Component
          key={filter.argName + filter.title}
          filterArgs={searchFilters}
          onChange={handleFilterChange}
        />
      ))}
      <td className="flex">
        <Button variant="outline-secondary" onClick={clearForm}>
          Clear
        </Button>
        <Button
          style={{marginLeft: '5px'}}
          variant="primary"
          onClick={() => onSubmit(searchFilters)}
        >
          Apply
        </Button>
      </td>
      <style jsx>{`
        .flex {
          display: flex;
        }
      `}</style>
    </tr>
  );
};

export default FilterableTableFilterRow;
