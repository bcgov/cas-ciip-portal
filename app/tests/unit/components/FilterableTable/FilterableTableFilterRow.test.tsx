import React from 'react';
import FilterableTableFilterRow from 'components/FilterableTable/FilterableTableFilterRow';
import {shallow, mount} from 'enzyme';
import {
  SortOnlyFilter,
  TableFilter,
  TextFilter
} from 'components/FilterableTable/Filters';

describe('The filterable table headers component', () => {
  it('renders search and reset buttons', () => {
    const rendered = shallow(
      <FilterableTableFilterRow
        filters={[]}
        onSubmit={jest.fn()}
        filterArgs={{}}
      />
    );
    expect(rendered).toMatchSnapshot();
    expect(
      rendered.find('Button[variant="outline-secondary"]').first().text()
    ).toBe('Clear');
    expect(rendered.find('Button[variant="primary"]').first().text()).toBe(
      'Apply'
    );
  });

  it('renders as many td elements as search options, plus one for the search buttons', () => {
    const filters: TableFilter[] = [
      new TextFilter('test', 'test'),
      new TextFilter('test2', 'test2')
    ];
    const rendered = mount(
      <FilterableTableFilterRow
        filters={filters}
        onSubmit={jest.fn()}
        filterArgs={{test: 'a', test2: 'b'}}
      />
    );

    expect(rendered).toMatchSnapshot();
    expect(rendered.find('td').length).toBe(3);
  });

  it('only renders elements that are searchable', () => {
    const filters = [
      new SortOnlyFilter('test', 'test'),
      new SortOnlyFilter('test2', 'test2'),
      new TextFilter('test3', 'test3')
    ];
    const rendered = mount(
      <FilterableTableFilterRow
        filters={filters}
        onSubmit={jest.fn()}
        filterArgs={{}}
      />
    );

    expect(rendered).toMatchSnapshot();
    expect(rendered.find('input').length).toBe(1);
    expect(rendered.find('input').first().props().name).toBe('test3');
  });

  it('should submit the filters if the "Enter" key is pressed', () => {
    const handleSubmit = jest.fn();
    const filters = [new TextFilter('test', 'test')];
    const rendered = shallow(
      <FilterableTableFilterRow
        filters={filters}
        onSubmit={handleSubmit}
        filterArgs={{}}
      />
    );

    rendered.find('tr').first().simulate('keyDown', {key: 'Enter'});

    expect(handleSubmit).toBeCalledTimes(1);
  });
});
