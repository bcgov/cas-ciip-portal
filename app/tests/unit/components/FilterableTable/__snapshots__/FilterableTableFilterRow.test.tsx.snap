// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The filterable table headers component only renders elements that are searchable 1`] = `
<tr
  className="jsx-870174952"
  onKeyDown={[Function]}
>
  <td
    className="jsx-870174952"
    key="testtest"
  />
  <td
    className="jsx-870174952"
    key="test2test2"
  />
  <td
    className="jsx-870174952"
    key="test3test3"
  >
    <FormControl
      aria-label="filter-by-test3"
      name="test3"
      onChange={[Function]}
      placeholder="Search"
      value=""
    />
  </td>
  <td
    className="jsx-870174952 flex"
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="outline-secondary"
    >
      Clear
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "5px",
        }
      }
      variant="primary"
    >
      Apply
    </Button>
  </td>
  <JSXStyle
    id="870174952"
  >
    .flex.jsx-870174952{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  </JSXStyle>
</tr>
`;

exports[`The filterable table headers component renders a option input if values are provided 1`] = `
<tr
  className="jsx-870174952"
  onKeyDown={[Function]}
>
  <td
    className="jsx-870174952"
    key="testtest"
  >
    <FormControl
      aria-label="filter-by-test"
      as="select"
      name="test"
      onChange={[Function]}
      placeholder="Search"
      value=""
    >
      <option
        className="jsx-870174952"
        key="test-placeholder"
        value=""
      >
        ...
      </option>
      <option
        className="jsx-870174952"
        key="test-option1"
        value="optionvalue1"
      >
        Option1
      </option>
      <option
        className="jsx-870174952"
        key="test-option2"
        value="optionvalue2"
      >
        Option2
      </option>
    </FormControl>
  </td>
  <td
    className="jsx-870174952 flex"
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="outline-secondary"
    >
      Clear
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "5px",
        }
      }
      variant="primary"
    >
      Apply
    </Button>
  </td>
  <JSXStyle
    id="870174952"
  >
    .flex.jsx-870174952{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  </JSXStyle>
</tr>
`;

exports[`The filterable table headers component renders a text input if values are not provided 1`] = `
<tr
  className="jsx-870174952"
  onKeyDown={[Function]}
>
  <td
    className="jsx-870174952"
    key="testtest"
  >
    <FormControl
      aria-label="filter-by-test"
      name="test"
      onChange={[Function]}
      placeholder="Search"
      value=""
    />
  </td>
  <td
    className="jsx-870174952 flex"
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="outline-secondary"
    >
      Clear
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "5px",
        }
      }
      variant="primary"
    >
      Apply
    </Button>
  </td>
  <JSXStyle
    id="870174952"
  >
    .flex.jsx-870174952{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  </JSXStyle>
</tr>
`;

exports[`The filterable table headers component renders as many td elements as search options, plus one for the search buttons 1`] = `
<tr
  className="jsx-870174952"
  onKeyDown={[Function]}
>
  <td
    className="jsx-870174952"
    key="testtest"
  >
    <FormControl
      aria-label="filter-by-test"
      name="test"
      onChange={[Function]}
      placeholder="Search"
      value="a"
    />
  </td>
  <td
    className="jsx-870174952"
    key="test2test2"
  >
    <FormControl
      aria-label="filter-by-test2"
      name="test2"
      onChange={[Function]}
      placeholder="Search"
      value="b"
    />
  </td>
  <td
    className="jsx-870174952 flex"
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="outline-secondary"
    >
      Clear
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "5px",
        }
      }
      variant="primary"
    >
      Apply
    </Button>
  </td>
  <JSXStyle
    id="870174952"
  >
    .flex.jsx-870174952{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  </JSXStyle>
</tr>
`;

exports[`The filterable table headers component renders search and reset buttons 1`] = `
<tr
  className="jsx-870174952"
  onKeyDown={[Function]}
>
  <td
    className="jsx-870174952 flex"
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="outline-secondary"
    >
      Clear
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "5px",
        }
      }
      variant="primary"
    >
      Apply
    </Button>
  </td>
  <JSXStyle
    id="870174952"
  >
    .flex.jsx-870174952{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
  </JSXStyle>
</tr>
`;
